File: processor/src/test/java/com/ioc/WeakDependencyTests.java
package com.ioc;

import com.google.testing.compile.JavaFileObjects;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.lang.ref.WeakReference;
import java.util.Arrays;

import javax.inject.Inject;
import javax.tools.JavaFileObject;

import static com.google.common.truth.Truth.assertAbout;
import static com.google.testing.compile.JavaSourcesSubjectFactory.javaSources;
import static com.ioc.Helpers.importType;

/**
 * Created by sergeygolishnikov on 01/11/2017.
 */
@RunWith(JUnit4.class)
public class WeakDependencyTests {
    @Test
    public void fieldInjectWithoutParams() throws Exception {
        JavaFileObject activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importType(Inject.class),
            importType(WeakReference.class),
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public WeakReference<DependencyModel> weakDependency;",
            "}");

        JavaFileObject dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            "public class DependencyModel {}");

        JavaFileObject injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            "import android.support.annotation.Keep",
            "import android.support.annotation.NonNull",
            importType(WeakReference.class),
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public final void inject(@NonNull final Activity target) {",
            "       injectDependencyModelInWeakDependency(target);",
            "   }",
            "",
            "   private final void injectDependencyModelInWeakDependency(@NonNull final Activity target) {",
            "       DependencyModel dependencyModel = new DependencyModel();",
            "       WeakReference<DependencyModel> weak_dependencyModel = new WeakReference<DependencyModel>(dependencyModel);",
            "       target.weakDependency = weak_dependencyModel;",
            "   }",
            "}");

        assertAbout(javaSources())
            .that(Arrays.asList(activityFile, dependencyFile))
            .processedWith(new IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile);
    }

    @Test
    public void fieldInjectWithParam() throws Exception {
        JavaFileObject activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importType(Inject.class),
            importType(WeakReference.class),
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public WeakReference<DependencyModel> weakDependency;",
            "}");

        JavaFileObject contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            importType(Dependency.class),
            "",
            "@Dependency",
            "public class Context {",
            "}");

        JavaFileObject dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importType(Inject.class),
            "",
            "public class DependencyModel {",
            "",
            "   @Inject",
            "   DependencyModel(Context context) {}",
            "}");

        JavaFileObject injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            "import android.support.annotation.Keep",
            "import android.support.annotation.NonNull",
            importType(WeakReference.class),
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public final void inject(@NonNull final Activity target) {",
            "       injectDependencyModelInWeakDependency(target);",
            "   }",
            "",
            "   private final void injectDependencyModelInWeakDependency(@NonNull final Activity target) {",
            "       Context context = new Context();",
            "       DependencyModel dependencyModel = new DependencyModel(context);",
            "       WeakReference<DependencyModel> weak_dependencyModel = new WeakReference<DependencyModel>(dependencyModel);",
            "       target.weakDependency = weak_dependencyModel;",
            "   }",
            "}");

        assertAbout(javaSources())
            .that(Arrays.asList(activityFile, contextFile, dependencyFile))
            .processedWith(new IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile);
    }


    @Test
    public void methodInjectWithoutParams() throws Exception {
        JavaFileObject activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importType(Inject.class),
            importType(WeakReference.class),
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   private WeakReference<DependencyModel> weakDependency;",
            "",
            "   public void weakDependency(WeakReference<DependencyModel> dependency) {};",
            "   public WeakReference<DependencyModel> getWeakDependency() { return null; };",
            "}");

        JavaFileObject dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            "public class DependencyModel {}");

        JavaFileObject injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            "import android.support.annotation.Keep",
            "import android.support.annotation.NonNull",
            importType(WeakReference.class),
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public final void inject(@NonNull final Activity target) {",
            "       injectDependencyModelInWeakDependency(target);",
            "   }",
            "",
            "   private final void injectDependencyModelInWeakDependency(@NonNull final Activity target) {",
            "       DependencyModel dependencyModel = new DependencyModel();",
            "       WeakReference<DependencyModel> weak_dependencyModel = new WeakReference<DependencyModel>(dependencyModel);",
            "       target.weakDependency(weak_dependencyModel);",
            "   }",
            "}");

        assertAbout(javaSources())
            .that(Arrays.asList(activityFile, dependencyFile))
            .processedWith(new IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile);
    }

    @Test
    public void methodInjectWithParam() throws Exception {
        JavaFileObject activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importType(Inject.class),
            importType(WeakReference.class),
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   private WeakReference<DependencyModel> weakDependency;",
            "",
            "   public void weakDependency(WeakReference<DependencyModel> dependency) {};",
            "   public WeakReference<DependencyModel> getWeakDependency() {return null;};",
            "}");

        JavaFileObject contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            importType(Dependency.class),
            "",
            "@Dependency",
            "public class Context {",
            "}");

        JavaFileObject dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importType(Inject.class),
            "",
            "public class DependencyModel {",
            "",
            "   @Inject",
            "   DependencyModel(Context context) {}",
            "}");

        JavaFileObject injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            "import android.support.annotation.Keep",
            "import android.support.annotation.NonNull",
            importType(WeakReference.class),
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public final void inject(@NonNull final Activity target) {",
            "       injectDependencyModelInWeakDependency(target);",
            "   }",
            "",
            "   private final void injectDependencyModelInWeakDependency(@NonNull final Activity target) {",
            "       Context context = new Context();",
            "       DependencyModel dependencyModel = new DependencyModel(context);",
            "       WeakReference<DependencyModel> weak_dependencyModel = new WeakReference<DependencyModel>(dependencyModel);",
            "       target.weakDependency(weak_dependencyModel);",
            "   }",
            "}");

        assertAbout(javaSources())
            .that(Arrays.asList(activityFile, contextFile, dependencyFile))
            .processedWith(new IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile);
    }


    @Test
    public void provideMethodInjectWithParam() throws Exception {
        JavaFileObject activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importType(Inject.class),
            importType(WeakReference.class),
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public DependencyModel dependency;",
            "}");

        JavaFileObject contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            "public class Context {",
            "}");

        JavaFileObject dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importType(Inject.class),
            "",
            "public class DependencyModel {",
            "",
            "   DependencyModel(Context context) {}",
            "}");

        JavaFileObject dependencyProvideModuleFile = JavaFileObjects.forSourceLines("test.DependencyProvide",
            "package test;",
            "",
            importType(Dependency.class),
            importType(WeakReference.class),
            "",
            "public class DependencyProvide {",
            "",
            "   @Dependency",
            "   public static DependencyModel getDependency(WeakReference<Context> context) { return null; }",
            "}");

        JavaFileObject injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            "import android.support.annotation.Keep",
            "import android.support.annotation.NonNull",
            importType(WeakReference.class),
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public final void inject(@NonNull final Activity target) {",
            "       injectDependencyModelInDependency(target);",
            "   }",
            "",
            "   private final void injectDependencyModelInDependency(@NonNull final Activity target) {",
            "       Context context = new Context();",
            "       WeakReference<Context> weak_context = new WeakReference<Context>(context);",
            "       DependencyModel dependencyModel = DependencyProvide.getDependency(weak_context);",
            "       target.dependency = dependencyModel;",
            "   }",
            "}");

        assertAbout(javaSources())
            .that(Arrays.asList(activityFile, dependencyProvideModuleFile, contextFile, dependencyFile))
            .processedWith(new IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile);
    }

    @Test
    public void provideWeakMethodInjectWithParam() throws Exception {
        JavaFileObject activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importType(Inject.class),
            importType(WeakReference.class),
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public DependencyModel dependency;",
            "}");

        JavaFileObject resourceFile = JavaFileObjects.forSourceLines("test.Resource",
            "package test;",
            "",
            "public class Resource {",
            "}");

        JavaFileObject contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            "public class Context {",
            "   Context(Resource resource) {}",
            "}");

        JavaFileObject dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importType(Inject.class),
            "",
            "public class DependencyModel {",
            "",
            "   DependencyModel(Context context) {}",
            "}");

        JavaFileObject dependencyProvideModuleFile = JavaFileObjects.forSourceLines("test.DependencyProvide",
            "package test;",
            "",
            importType(Dependency.class),
            importType(WeakReference.class),
            "",
            "public class DependencyProvide {",
            "",
            "   @Dependency",
            "   public static DependencyModel getDependency(WeakReference<Context> context) { return null; }",
            "}");

        JavaFileObject injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            "import android.support.annotation.Keep",
            "import android.support.annotation.NonNull",
            importType(WeakReference.class),
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public final void inject(@NonNull final Activity target) {",
            "       injectDependencyModelInDependency(target);",
            "   }",
            "",
            "   private final void injectDependencyModelInDependency(@NonNull final Activity target) {",
            "       Resource resource = new Resource();",
            "       Context context = new Context(resource);",
            "       WeakReference<Context> weak_context = new WeakReference<Context>(context);",
            "       DependencyModel dependencyModel = DependencyProvide.getDependency(weak_context);",
            "       target.dependency = dependencyModel;",
            "   }",
            "}");

        assertAbout(javaSources())
            .that(Arrays.asList(activityFile, resourceFile, dependencyProvideModuleFile, contextFile, dependencyFile))
            .processedWith(new IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile);
    }

    @Test
    public void provideMethodInject() throws Exception {
        JavaFileObject activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importType(Inject.class),
            importType(WeakReference.class),
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public WeakReference<DependencyModel> dependency;",
            "}");

        JavaFileObject dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importType(Inject.class),
            "",
            "public class DependencyModel {",
            "",
            "   DependencyModel() {}",
            "}");

        JavaFileObject dependencyProvideModuleFile = JavaFileObjects.forSourceLines("test.DependencyProvide",
            "package test;",
            "",
            importType(Dependency.class),
            importType(WeakReference.class),
            "",
            "public class DependencyProvide {",
            "",
            "   @Dependency",
            "   public static DependencyModel getDependency() { return null; }",
            "}");

        JavaFileObject injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            "import android.support.annotation.Keep",
            "import android.support.annotation.NonNull",
            importType(WeakReference.class),
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public final void inject(@NonNull final Activity target) {",
            "       injectDependencyModelInDependency(target);",
            "   }",
            "",
            "   private final void injectDependencyModelInDependency(@NonNull final Activity target) {",
            "       DependencyModel dependencyModel = DependencyProvide.getDependency();",
            "       WeakReference<DependencyModel> weak_dependencyModel = new WeakReference<DependencyModel>(dependencyModel);",
            "       target.dependency = weak_dependencyModel;",
            "   }",
            "}");

        assertAbout(javaSources())
            .that(Arrays.asList(activityFile, dependencyProvideModuleFile, dependencyFile))
            .processedWith(new IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile);
    }

    @Test
    public void provideMethodInjectWithParamWeak() throws Exception {
        JavaFileObject activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importType(Inject.class),
            importType(WeakReference.class),
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public WeakReference<DependencyModel> dependency;",
            "}");

        JavaFileObject contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            "public class Context {",
            "}");

        JavaFileObject dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importType(Inject.class),
            "",
            "public class DependencyModel {",
            "",
            "   DependencyModel(Context context) {}",
            "}");

        JavaFileObject dependencyProvideModuleFile = JavaFileObjects.forSourceLines("test.DependencyProvide",
            "package test;",
            "",
            importType(Dependency.class),
            importType(WeakReference.class),
            "",
            "public class DependencyProvide {",
            "",
            "   @Dependency",
            "   public static DependencyModel getDependency(Context context) { return null; }",
            "}");

        JavaFileObject injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            "import android.support.annotation.Keep",
            "import android.support.annotation.NonNull",
            importType(WeakReference.class),
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public final void inject(@NonNull final Activity target) {",
            "       injectDependencyModelInDependency(target);",
            "   }",
            "",
            "   private final void injectDependencyModelInDependency(@NonNull final Activity target) {",
            "       Context context = new Context();",
            "       DependencyModel dependencyModel = DependencyProvide.getDependency(context);",
            "       WeakReference<DependencyModel> weak_dependencyModel = new WeakReference<DependencyModel>(dependencyModel);",
            "       target.dependency = weak_dependencyModel;",
            "   }",
            "}");

        assertAbout(javaSources())
            .that(Arrays.asList(activityFile, dependencyProvideModuleFile, contextFile, dependencyFile))
            .processedWith(new IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile);
    }

    @Test
    public void fieldInjectWithConstructorParamWeak() throws Exception {
        JavaFileObject activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importType(Inject.class),
            importType(WeakReference.class),
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public DependencyModel dependency;",
            "}");

        JavaFileObject contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            importType(Dependency.class),
            "",
            "@Dependency",
            "public class Context {",
            "}");

        JavaFileObject dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importType(Inject.class),
            importType(WeakReference.class),
            "",
            "public class DependencyModel {",
            "",
            "   @Inject",
            "   DependencyModel(WeakReference<Context> context) {}",
            "}");

        JavaFileObject injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            "import android.support.annotation.Keep",
            "import android.support.annotation.NonNull",
            importType(WeakReference.class),
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public final void inject(@NonNull final Activity target) {",
            "       injectDependencyModelInDependency(target);",
            "   }",
            "",
            "   private final void injectDependencyModelInDependency(@NonNull final Activity target) {",
            "       Context context = new Context();",
            "       WeakReference<Context> weak_context = new WeakReference<Context>(context);",
            "       DependencyModel dependencyModel = new DependencyModel(weak_context);",
            "       target.dependency = dependencyModel;",
            "   }",
            "}");

        assertAbout(javaSources())
            .that(Arrays.asList(activityFile, contextFile, dependencyFile))
            .processedWith(new IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile);
    }

}

File: processor/src/test/java/com/ioc/WeakDependencyTests.kt
package com.ioc

import com.google.common.truth.Truth.assertAbout
import com.google.testing.compile.JavaFileObjects
import com.google.testing.compile.JavaSourcesSubject
import com.google.testing.compile.JavaSourcesSubjectFactory.javaSources
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import javax.tools.JavaFileObject

/**
 * Created by sergeygolishnikov on 01/11/2017.
 */
@RunWith(JUnit4::class)
class WeakDependencyTests {
    @Test
    @Throws(Exception::class)
    fun fieldInjectWithoutParams() {
        val activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importInjectAnnotation,
            importWeakReference,
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public WeakReference<DependencyModel> weakDependency;",
            "}")

        val dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            "public class DependencyModel {}")

        val injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            importKeepAnnotation,
            importNonNullAnnotation,
            "import $weakReferenceType",
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public static final void inject(@NonNull final Activity target) {",
            "       target.weakDependency = new WeakReference<>(new DependencyModel());",
            "   }",
            "}")

        assertAbout<JavaSourcesSubject, Iterable<JavaFileObject>>(javaSources())
            .that(listOf(activityFile, dependencyFile))
            .processedWith(IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile)
    }

    @Test
    @Throws(Exception::class)
    fun fieldInjectWithParam() {
        val activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importInjectAnnotation,
            importWeakReference,
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public WeakReference<DependencyModel> weakDependency;",
            "}")

        val contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            importDependencyAnnotation,
            "",
            "@Dependency",
            "public class Context {",
            "}")

        val dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importInjectAnnotation,
            "",
            "public class DependencyModel {",
            "",
            "   @Inject",
            "   DependencyModel(Context context) {}",
            "}")

        val injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            importKeepAnnotation,
            importNonNullAnnotation,
            importWeakReference,
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public static final void inject(@NonNull final Activity target) {",
            "       target.weakDependency = new WeakReference<>(provideDependencyModel());",
            "   }",
            "",
            "   public static final DependencyModel provideDependencyModel() {",
            "       Context context = new Context();",
            "       DependencyModel dependencyModel = new DependencyModel(context);",
            "       return dependencyModel;",
            "   }",
            "}")

        assertAbout<JavaSourcesSubject, Iterable<JavaFileObject>>(javaSources())
            .that(listOf(activityFile, contextFile, dependencyFile))
            .processedWith(IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile)
    }


    @Test
    @Throws(Exception::class)
    fun methodInjectWithoutParams() {
        val activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importInjectAnnotation,
            importWeakReference,
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   private WeakReference<DependencyModel> weakDependency;",
            "",
            "   public void weakDependency(WeakReference<DependencyModel> dependency) {};",
            "   public WeakReference<DependencyModel> getWeakDependency() { return null; };",
            "}")

        val dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            "public class DependencyModel {}")

        val injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            importKeepAnnotation,
            importNonNullAnnotation,
            importWeakReference,
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public static final void inject(@NonNull final Activity target) {",
            "       target.weakDependency(new WeakReference<>(new DependencyModel()));",
            "   }",
            "}")

        assertAbout<JavaSourcesSubject, Iterable<JavaFileObject>>(javaSources())
            .that(listOf(activityFile, dependencyFile))
            .processedWith(IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile)
    }

    @Test
    @Throws(Exception::class)
    fun methodInjectWithParam() {
        val activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importInjectAnnotation,
            importWeakReference,
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   private WeakReference<DependencyModel> weakDependency;",
            "",
            "   public void weakDependency(WeakReference<DependencyModel> dependency) {};",
            "   public WeakReference<DependencyModel> getWeakDependency() {return null;};",
            "}")

        val contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            importDependencyAnnotation,
            "",
            "@Dependency",
            "public class Context {",
            "}")

        val dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importInjectAnnotation,
            "",
            "public class DependencyModel {",
            "",
            "   @Inject",
            "   DependencyModel(Context context) {}",
            "}")

        val injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            importKeepAnnotation,
            importNonNullAnnotation,
            importWeakReference,
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public static final void inject(@NonNull final Activity target) {",
            "       target.weakDependency(new WeakReference<>(provideDependencyModel()));",
            "   }",
            "",
            "   public static final DependencyModel provideDependencyModel() {",
            "       Context context = new Context();",
            "       DependencyModel dependencyModel = new DependencyModel(context);",
            "       return dependencyModel;",
            "   }",
            "}")

        assertAbout<JavaSourcesSubject, Iterable<JavaFileObject>>(javaSources())
            .that(listOf(activityFile, contextFile, dependencyFile))
            .processedWith(IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile)
    }


    @Test
    @Throws(Exception::class)
    fun provideMethodInjectWithParam() {
        val activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importInjectAnnotation,
            importWeakReference,
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public DependencyModel dependency;",
            "}")

        val contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            "public class Context {",
            "}")

        val dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importInjectAnnotation,
            "",
            "public class DependencyModel {",
            "",
            "   DependencyModel(Context context) {}",
            "}")

        val dependencyProvideModuleFile = JavaFileObjects.forSourceLines("test.DependencyProvide",
            "package test;",
            "",
            importDependencyAnnotation,
            importWeakReference,
            "",
            "public class DependencyProvide {",
            "",
            "   @Dependency",
            "   public static DependencyModel getDependency(WeakReference<Context> context) { return null; }",
            "}")

        val injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            importKeepAnnotation,
            importNonNullAnnotation,
            importWeakReference,
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public static final void inject(@NonNull final Activity target) {",
            "       target.dependency = provideDependencyModel();",
            "   }",
            "",
            "   public static final DependencyModel provideDependencyModel() {",
            "       Context context = new Context();",
            "       WeakReference<Context> weakContext = new WeakReference<Context>(context);",
            "       DependencyModel dependencyModel = DependencyProvide.getDependency(weakContext);",
            "       return dependencyModel;",
            "   }",
            "}")

        assertAbout<JavaSourcesSubject, Iterable<JavaFileObject>>(javaSources())
            .that(listOf(activityFile, dependencyProvideModuleFile, contextFile, dependencyFile))
            .processedWith(IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile)
    }

    @Test
    @Throws(Exception::class)
    fun provideWeakMethodInjectWithParam() {
        val activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importInjectAnnotation,
            importWeakReference,
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public DependencyModel dependency;",
            "}")

        val resourceFile = JavaFileObjects.forSourceLines("test.Resource",
            "package test;",
            "",
            "public class Resource {",
            "}")

        val contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            "public class Context {",
            "   Context(Resource resource) {}",
            "}")

        val dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importInjectAnnotation,
            "",
            "public class DependencyModel {",
            "",
            "   DependencyModel(Context context) {}",
            "}")

        val dependencyProvideModuleFile = JavaFileObjects.forSourceLines("test.DependencyProvide",
            "package test;",
            "",
            importDependencyAnnotation,
            importWeakReference,
            "",
            "public class DependencyProvide {",
            "",
            "   @Dependency",
            "   public static DependencyModel getDependency(WeakReference<Context> context) { return null; }",
            "}")

        val injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            importKeepAnnotation,
            importNonNullAnnotation,
            importWeakReference,
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public static final void inject(@NonNull final Activity target) {",
            "       target.dependency = provideDependencyModel();",
            "   }",
            "",
            "   public static final DependencyModel provideDependencyModel() {",
            "       Resource resource = new Resource();",
            "       Context context = new Context(resource);",
            "       WeakReference<Context> weakContext = new WeakReference<Context>(context);",
            "       DependencyModel dependencyModel = DependencyProvide.getDependency(weakContext);",
            "       return dependencyModel;",
            "   }",
            "}")

        assertAbout<JavaSourcesSubject, Iterable<JavaFileObject>>(javaSources())
            .that(listOf(activityFile, resourceFile, dependencyProvideModuleFile, contextFile, dependencyFile))
            .processedWith(IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile)
    }

    @Test
    @Throws(Exception::class)
    fun provideMethodInject() {
        val activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importInjectAnnotation,
            importWeakReference,
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public WeakReference<DependencyModel> dependency;",
            "}")

        val dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importInjectAnnotation,
            "",
            "public class DependencyModel {",
            "",
            "   DependencyModel() {}",
            "}")

        val dependencyProvideModuleFile = JavaFileObjects.forSourceLines("test.DependencyProvide",
            "package test;",
            "",
            importDependencyAnnotation,
            importWeakReference,
            "",
            "public class DependencyProvide {",
            "",
            "   @Dependency",
            "   public static DependencyModel getDependency() { return null; }",
            "}")

        val injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            importKeepAnnotation,
            importNonNullAnnotation,
            importWeakReference,
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public static final void inject(@NonNull final Activity target) {",
            "       target.dependency = new WeakReference<>(DependencyProvide.getDependency());",
            "   }",
            "}")

        assertAbout<JavaSourcesSubject, Iterable<JavaFileObject>>(javaSources())
            .that(listOf(activityFile, dependencyProvideModuleFile, dependencyFile))
            .processedWith(IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile)
    }

    @Test
    @Throws(Exception::class)
    fun provideMethodInjectWithParamWeak() {
        val activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importInjectAnnotation,
            importWeakReference,
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public WeakReference<DependencyModel> dependency;",
            "}")

        val contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            "public class Context {",
            "}")

        val dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importInjectAnnotation,
            "",
            "public class DependencyModel {",
            "",
            "   DependencyModel(Context context) {}",
            "}")

        val dependencyProvideModuleFile = JavaFileObjects.forSourceLines("test.DependencyProvide",
            "package test;",
            "",
            importDependencyAnnotation,
            importWeakReference,
            "",
            "public class DependencyProvide {",
            "",
            "   @Dependency",
            "   public static DependencyModel getDependency(Context context) { return null; }",
            "}")

        val injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            importKeepAnnotation,
            importNonNullAnnotation,
            importWeakReference,
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public static final void inject(@NonNull final Activity target) {",
            "       target.dependency = new WeakReference<>(provideDependencyModel());",
            "   }",
            "",
            "   public static final DependencyModel provideDependencyModel() {",
            "       Context context = new Context();",
            "       DependencyModel dependencyModel = DependencyProvide.getDependency(context);",
            "       return dependencyModel;",
            "   }",
            "}")

        assertAbout<JavaSourcesSubject, Iterable<JavaFileObject>>(javaSources())
            .that(listOf(activityFile, dependencyProvideModuleFile, contextFile, dependencyFile))
            .processedWith(IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile)
    }

    @Test
    @Throws(Exception::class)
    fun fieldInjectWithConstructorParamWeak() {
        val activityFile = JavaFileObjects.forSourceLines("test.Activity",
            "package test;",
            "",
            importInjectAnnotation,
            importWeakReference,
            "",
            "public class Activity {",
            "",
            "   @Inject",
            "   public DependencyModel dependency;",
            "}")

        val contextFile = JavaFileObjects.forSourceLines("test.Context",
            "package test;",
            "",
            importDependencyAnnotation,
            "",
            "@Dependency",
            "public class Context {",
            "}")

        val dependencyFile = JavaFileObjects.forSourceLines("test.DependencyModel",
            "package test;",
            "",
            importInjectAnnotation,
            importWeakReference,
            "",
            "public class DependencyModel {",
            "",
            "   @Inject",
            "   DependencyModel(WeakReference<Context> context) {}",
            "}")

        val injectedFile = JavaFileObjects.forSourceLines("test.ActivityInjector",
            "package test;",
            "",
            importKeepAnnotation,
            importNonNullAnnotation,
            importWeakReference,
            "",
            "@Keep",
            "public final class ActivityInjector {",
            "   @Keep",
            "   public static final void inject(@NonNull final Activity target) {",
            "       target.dependency = provideDependencyModel();",
            "   }",
            "",
            "   public static final DependencyModel provideDependencyModel() {",
            "       Context context = new Context();",
            "       WeakReference<Context> weakContext = new WeakReference<Context>(context);",
            "       DependencyModel dependencyModel = new DependencyModel(weakContext);",
            "       return dependencyModel;",
            "   }",
            "}")

        assertAbout<JavaSourcesSubject, Iterable<JavaFileObject>>(javaSources())
            .that(listOf(activityFile, contextFile, dependencyFile))
            .processedWith(IProcessor())
            .compilesWithoutError()
            .and().generatesSources(injectedFile)
    }
}
